### PR TITLE
Fix: avoid `object-shorthand` crash with spread properties (fixes #7305)

### LINKED ISSUE #7305: object-shorthand consistent-as-needed crashes linting after last object-shorthand rule update
**Tell us about your environment**

* **ESLint Version:** 3.7.1
* **Node Version:** 5.12.0
* **npm Version:** 3.8.6

**What parser (default, Babel-ESLint, etc.) are you using?** default

**Please show your full configuration:**
```json
{
  "plugins": [],
  "env": {
    "browser": true,
    "es6": true,
    "node": true
  },
  "parserOptions": {
    "sourceType": "module",
    "impliedStrict": true,
    "ecmaVersion": 6,
    "ecmaFeatures": {
      "experimentalObjectRestSpread": true
    }
  },
  "rules": {
    "object-shorthand": ["error", "consistent-as-needed"]
  },
  "globals": {}
}
```

**What did you do? Please include the actual source code causing the issue.**

```js
const x = {};
const y = { ... x };
```

**What did you expect to happen?**
Eslint should finish its job of linting specified files.

**What actually happened? Please include the actual, raw output from ESLint.**
```
Cannot read property 'type' of undefined
TypeError: Cannot read property 'type' of undefined
    at isRedundant (/home/dzek/Projects/eslint-fail/node_modules/eslint/lib/rules/object-shorthand.js:160:22)
    at Array.every (native)
    at checkConsistency (/home/dzek/Projects/eslint-fail/node_modules/eslint/lib/rules/object-shorthand.js:196:66)
    at EventEmitter.ObjectExpression (/home/dzek/Projects/eslint-fail/node_modules/eslint/lib/rules/object-shorthand.js:215:21)
    at emitOne (events.js:90:13)
    at EventEmitter.emit (events.js:182:7)
    at NodeEventGenerator.enterNode (/home/dzek/Projects/eslint-fail/node_modules/eslint/lib/util/node-event-generator.js:40:22)
    at CodePathAnalyzer.enterNode (/home/dzek/Projects/eslint-fail/node_modules/eslint/lib/code-path-analysis/code-path-analyzer.js:608:23)
    at CommentEventGenerator.enterNode (/home/dzek/Projects/eslint-fail/node_modules/eslint/lib/util/comment-event-generator.js:97:23)
    at Controller.enter (/home/dzek/Projects/eslint-fail/node_modules/eslint/lib/eslint.js:898:36)
```

### PR DESCRIPTION
**Pull request by not-an-aardvark**
**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[x] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

See https://github.com/eslint/eslint/issues/7305

**Please check each item to ensure your pull request is ready:**

- [x] I've read the [pull request guide](http://eslint.org/docs/developer-guide/contributing/pull-requests)
- [x] I've included tests for my change
- (n/a) I've updated documentation for my change (if appropriate)

**What changes did you make? (Give an overview)**

This updates `object-shorthand` to work correctly with spread properties. Previously, it would crash when it encountered them with the `consistent-as-needed` option; now it categorizes them with getters and setters as properties that do not have a shorthand form.

**Is there anything you'd like reviewers to focus on?**

Nothing in particular.

### CODE REVIEW

File: tests/lib/rules/object-shorthand.js [hunk 1]
```diff
@@ -109,7 +109,8 @@ ruleTester.run("object-shorthand", rule, {
         { code: "var x = {foo: 'foo'}", parserOptions: { ecmaVersion: 6}, options: ["consistent-as-needed"] },
         { code: "var x = {[foo]: foo}", parserOptions: { ecmaVersion: 6}, options: ["consistent-as-needed"] },
         { code: "var x = {foo: function foo() {}}", parserOptions: { ecmaVersion: 6}, options: ["consistent-as-needed"] },
-        { code: "var x = {[foo]: 'foo'}", parserOptions: { ecmaVersion: 6}, options: ["consistent-as-needed"] }
+        { code: "var x = {[foo]: 'foo'}", parserOptions: { ecmaVersion: 6}, options: ["consistent-as-needed"] },
+        { code: "var foo = {...bar}", parserOptions: { ecmaVersion: 6, ecmaFeatures: { experimentalObjectRestSpread: true } }, options: ["consistent-as-needed"] }
```
Comment (gyandeeps) [inline tests/lib/rules/object-shorthand.js]: I would recommend adding more tests for different rule options in combination with object spread.

Example:

```js
var foo = {...bar, a, b};

var foo = {...bar, a: 'foo', b: 'bar'};
```

File: lib/rules/object-shorthand.js [hunk 2]
```diff
@@ -176,7 +176,7 @@ module.exports = {
         function checkConsistency(node, checkRedundancy) {
 
             // We are excluding getters and setters as they are considered neither longform nor shorthand.
```
Comment (kaicataldo) [inline lib/rules/object-shorthand.js]: Can we update this comment to reflect the new changes?

Comment (not-an-aardvark) [inline lib/rules/object-shorthand.js]: Updated.

File: tests/lib/rules/object-shorthand.js (hunk 1 above)
Comment (not-an-aardvark) [inline tests/lib/rules/object-shorthand.js]: Added some more tests.

File: tests/lib/rules/object-shorthand.js [hunk 3]
```diff
@@ -161,10 +172,13 @@ ruleTester.run("object-shorthand", rule, {
...
 
         // consistent-as-needed
         { code: "var x = {a: a, b: b}", parserOptions: { ecmaVersion: 6}, errors: [{ message: "Expected shorthand for all properties.", type: "" }], options: ["consistent-as-needed"] },
         { code: "var x = {a, z: function z(){}}", parserOptions: { ecmaVersion: 6}, errors: [{ message: "Unexpected mix of shorthand and non-shorthand properties.", type: "" }], options: ["consistent-as-needed"] },
         { code: "var x = {foo: function() {}}", parserOptions: { ecmaVersion: 6}, errors: [{ message: "Expected shorthand for all properties.", type: "" }], options: ["consistent-as-needed"] },
+        { code: "var x = {a: a, b: b, ...baz}", parserOptions: { ecmaVersion: 6, ecmaFeatures: { experimentalObjectRestSpread: true } }, errors: [{ message: "Expected shorthand for all properties.", type: "" }], options: ["consistent-as-needed"] },
+        { code: "var x = {foo, bar: bar, ...qux}", parserOptions: { ecmaVersion: 6, ecmaFeatures: { experimentalObjectRestSpread: true } }, errors: [{ message: "Unexpected mix of shorthand and non-shorthand properties.", type: "" }], options: ["consistent-as-needed"] }
```
Comment (not-an-aardvark) [inline tests/lib/rules/object-shorthand.js]: This is only loosely related to this PR. but in my opinion this file is a bit difficult to read. All of the test cases are on a line and are closely packed together, so it makes the file feel very dense.

If other people feel the same way, maybe we should add some linebreaks (presumably in a different PR).

Comment (nzakas) [inline tests/lib/rules/object-shorthand.js]: Feel free to open a new PR for that. We should keep this one narrowly focused

**Pull request merged**